### PR TITLE
1014 learning rate finder

### PR DESCRIPTION
**Pull request by rijobro**
Fixes https://github.com/Project-MONAI/MONAI/issues/1014.

### Description
Implements calculation of optimal learning rate based on https://github.com/davidtvs/pytorch-lr-finder.

image

### Status
**Ready**

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [x] New tests added to cover the changes.
- [x] Integration tests passed locally by running `./runtests.sh --codeformat --coverage`.
- [x] Quick tests passed locally by running `./runtests.sh --quick`.
- [x] In-line docstrings updated.
- [x] Documentation updated, tested `make html` command in the `docs/` folder.

### CODE REVIEW

**Comment by rijobro** [comment]
@wyli @Nic-Ma @ericspod [this part of the PR](https://github.com/Project-MONAI/MONAI/pull/1454/files#diff-227861f2b9c5eec7c60d554b7c3b52cb2a642fdeb7adaf648789785bdd9d11b3R587) saves the state of the network and optimiser to disk or memory, such that they can be restored at the end. 

Does this functionality live somewhere else in MONAI?

**Comment by ericspod** [comment]
We had that kind of functionality as part of an [Ignite handler for saving checkpoints](https://github.com/Project-MONAI/MONAI/blob/master/monai/handlers/checkpoint_saver.py), but that's not quite how you're doing things here.

File: monai/optimizers/lr_finder.py [hunk 1]
```diff
@@ -0,0 +1,629 @@
...
+    """Exponentially increases the learning rate between two boundaries over a number of
+    iterations.
+    """
+
+    def get_lr(self):
+        r = self.last_epoch / (self.num_iter - 1)
+        return [base_lr * (self.end_lr / base_lr) ** r for base_lr in self.base_lrs]
+
+
+class StateCacher(object):
```
Comment (wyli) [inline monai/optimizers/lr_finder.py:587]: I feel this is useful as a general utility, perhaps move it to monai.utils and add some docstrings? perhaps `state_dict` could be `data_obj`? this class doesn't assume a specific `state`

File: monai/optimizers/lr_finder.py [hunk 2]
```diff
@@ -0,0 +1,629 @@
...
+            self.cached.update({key: fn})
+            torch.save(state_dict, fn)
+
+    def retrieve(self, key):
+        if key not in self.cached:
+            raise KeyError(f"Target {key} was not cached.")
+
+        if self.in_memory:
+            return self.cached.get(key)
+        else:
```
Comment (wyli) [inline monai/optimizers/lr_finder.py:616]: this `else:` is not necessary because line 615 already returned

File: monai/optimizers/lr_finder.py [hunk 3]
```diff
@@ -0,0 +1,629 @@
...
+            ax: the plot is created in the specified matplotlib axes object and the
+                figure is not be shown. If `None`, then the figure and axes object are
+                created in this method and the figure is shown.
+            steepest_lr: plot the learning rate which had the steepest gradient.
+
+        Returns:
+            The matplotlib.axes.Axes object that contains the plot
+        """
+        if not has_matplotlib:
+            raise RuntimeError("Matplotlib is missing, can't plot result")
```
Comment (wyli) [inline monai/optimizers/lr_finder.py:512]: could be a warning and return None?

File: monai/optimizers/lr_finder.py [hunk 4]
```diff
@@ -0,0 +1,629 @@
...
+            total_loss += loss.item()
+
+        self.optimizer.step()
+
+        return total_loss
+
+    def _move_to_device(
+        self, inputs: torch.Tensor, labels: torch.Tensor, non_blocking: bool = True
+    ) -> Tuple[torch.Tensor, torch.Tensor]:
+        def move(obj, device, non_blocking=True):
```
Comment (wyli) [inline monai/optimizers/lr_finder.py:410]: this  could be a generic `to_device` util in monai.utils, but it'll need some recursive logic like this one: https://github.com/pytorch/pytorch/blob/master/torch/utils/data/_utils/collate.py#L15

**Review by wyli — COMMENTED** [review]
thanks @rijobro this is very useful! 

Perhaps the way the training losses is computed/accumulated (`self._train_batch`) should be decoupled from the lr_finder. Intuitively I think the LR finder should work fine as long as the user provides a black box that takes lr/step as input, and returns a `total_loss` (e.g. `self._train_batch` and `self._validate`). Do you want to refactor in this PR to decouple those? otherwise we could file a ticket and have another iteration.

pls see also some minor suggestions inline, they are mostly optional

@Can-Zhao would be great to have your comments as well!

File: monai/optimizers/lr_finder.py [hunk 5]
```diff
@@ -0,0 +1,629 @@
...
+
+    has_tqdm = True
+except ImportError:
+    has_tqdm = False
+try:
+    import matplotlib.pyplot as plt
+
+    has_matplotlib = True
+except ImportError:
+    has_matplotlib = False
```
[the remark this answers is not in the document]
Comment (rijobro) [reply monai/optimizers/lr_finder.py:27]: I had this originall, but it was causing the CI to fail. I'll switch back.

**Comment by rijobro** [comment]
@wyli thanks, I'll get to it!

**Comment by rijobro** [comment]
@wyli this is ready if you want to review again, thanks!

**Review by wyli — APPROVED** [review]
thanks, it looks good, except some minor warnings on `(object)` https://deepsource.io/gh/Project-MONAI/MONAI/run/a53a8313-ab56-4aea-8a8b-e6fd4941f978/python/PYL-R0205 this new feature needs another iteration to decouple the actual training/validation logic`_train_batch` and `_validate` from the `LearningRateFinder` class

**Comment by ShadowTwin41** [comment]
I am having trouble using this class for GANs. Is there any way to do something similar to the "GANLearner" fastai?
